File: AdaptationforFoilingCatamaran/computeResidual.m
function residual = computeResidual(x,eta,nu,foil,wind)
% eta = [0;0;x(3);x(4);x(5);0];
% nu = [[x(1)*cos(x(2));x(1)*sin(x(2));0]; zeros(3,1)];
% nu = [Rbn(eta).'* [x(1)*cos(x(2));x(1)*sin(x(2));0]; zeros(3,1)];

% Adjust L foil
foil{2}.attitudeInB = foil{2}.attitudeInB + [0;x(1);0];
foil{3}.attitudeInB = foil{3}.attitudeInB + [0;x(1);0];

% Adjust Starboard T foil
foil{4}.attitudeInB = foil{4}.attitudeInB + [0;x(2);x(4)]; 
% Adjust Starboard rudder
foil{5}.attitudeInB = foil{5}.attitudeInB + [0;0;x(4)]; 

% Adjust Port T foil
foil{6}.attitudeInB = foil{6}.attitudeInB + [0;x(3);x(4)]; 
% Adjust Port rudder
foil{7}.attitudeInB = foil{7}.attitudeInB + [0;0;x(4)]; 

totalLoad = zeros(6,1);
for idxFoil=1:length(foil)
totalLoad = totalLoad + foilLoad(eta,nu,foil{idxFoil},wind,[],false);
end
totalLoad= totalLoad + weightLoad(eta,false);
totalLoad = totalLoad + aerodynamicLoadSuperstructure(eta,nu,wind,false);

w1 = 1;
w2 = 1;
residual = w1*sqrt(sum((1e-3*totalLoad(1:3)).^2)) + w2*sqrt(sum((1e-3*totalLoad(4:6)).^2));
end

File: rlController/computeResidual.py
import numpy as np
from weightLoad import weightLoad
from aerodynamicLoadSuperstructure import aerodynamicLoadSuperstructure
from foilLoad import foilLoad

def computeResidual(x, eta, nu, foil, wind):
    """
    x : optimization variables
    eta, nu : state vectors
    foil : list of foil dicts
    wind : wind dict
    """

    # --------------------------------------------------
    # Copy foil attitudes so we don't modify them in-place
    # --------------------------------------------------
    foil_local = []
    for f in foil:
        f_copy = f.copy()
        f_copy["attitudeInB"] = f["attitudeInB"].copy()
        foil_local.append(f_copy)

    # --------------------------------------------------
    # Adjust L foils (starboard horizontal + vertical)
    # MATLAB: foil{2}, foil{3}
    # Python: foil_local[1], foil_local[2]
    # --------------------------------------------------
    foil_local[1]["attitudeInB"] += np.array([0.0, x[1], 0.0])
    foil_local[2]["attitudeInB"] += np.array([0.0, x[1], 0.0])

    # --------------------------------------------------
    # Adjust starboard T foil and rudder
    # MATLAB: foil{4}, foil{5}
    # Python: foil_local[3], foil_local[4]
    # --------------------------------------------------
    foil_local[3]["attitudeInB"] += np.array([0.0, x[2], x[4]])
    foil_local[4]["attitudeInB"] += np.array([0.0, 0.0, x[4]])

    # --------------------------------------------------
    # Adjust port T foil and rudder
    # MATLAB: foil{6}, foil{7}
    # Python: foil_local[5], foil_local[6]
    # --------------------------------------------------
    foil_local[5]["attitudeInB"] += np.array([0.0, x[3], x[4]])
    foil_local[6]["attitudeInB"] += np.array([0.0, 0.0, x[4]])

    # --------------------------------------------------
    # Sum all loads
    # --------------------------------------------------
    totalLoad = np.zeros(6)

    for f in foil_local:
        totalLoad += foilLoad(eta, nu, f, wind, [], False)

    totalLoad += weightLoad(eta, False)
    totalLoad += aerodynamicLoadSuperstructure(eta, nu, wind, False)

    # --------------------------------------------------
    # Residual
    # --------------------------------------------------
    w1 = 1.0
    w2 = 1.0

    force_residual = np.sqrt(1e-3 * np.sum(totalLoad[0:3]**2))
    moment_residual = np.sqrt(1e-3 * np.sum(totalLoad[3:6]**2))

    residual = w1 * force_residual + w2 * moment_residual

    return residual
